include/nuttx/compiler.h

```diff
@@ -69,6 +69,36 @@
 #  undef CONFIG_HAVE_CXX14
 #endif
 
+#if defined(__cplusplus) && __cplusplus >= 201703L
+#  define CONFIG_HAVE_CXX17 1
+#else
+#  undef CONFIG_HAVE_CXX17
+#endif
+
+
+/* Check for intrinsics */
+#ifdef __has_builtin
+# if  __has_builtin(__builtin_bswap16)
+#  define CONFIG_HAVE_BUILTIN_BSWAP16 1
+# endif
+# if  __has_builtin(__builtin_bswap32)
+#  define CONFIG_HAVE_BUILTIN_BSWAP32 1
+# endif
+# if  __has_builtin(__builtin_bswap64) && defined(CONFIG_HAVE_LONG_LONG)
+#  define CONFIG_HAVE_BUILTIN_BSWAP64 1
+# endif
+# if  __has_builtin(__builtin_ctz)
+#  define CONFIG_HAVE_BUILTIN_CTZ 1
+# endif
+# if  __has_builtin(__builtin_clz)
+#  define CONFIG_HAVE_BUILTIN_CLZ 1
+# endif
+# if  __has_builtin(__builtin_popcount)
+#  define CONFIG_HAVE_BUILTIN_POPCOUNT 1
+# endif
+#endif
```

Comment:
Please move all this block under `#ifdef __GNUC__`.
Also according https://stackoverflow.com/questions/54079257/how-to-check-builtin-function-is-available-on-gcc the `__has_builtin` is a new GCC feature while the old version might still have builtin functions available, but will fail `__has_builtin` check. It is better to rely on the GCC version.

Reply:
I just switched these to checking the symbol itself, since `__GNUC__` can be disabled on clang but the builtins might still exist.

Reply:
I think Clag also defines `__GNUC__`

Reply:
done. Clang does unless you disable it and it may not keep up with GCC's one

Reply:
Yes. For Clang you can even redefine `__GNUC__` to a desired value, so with compiler option you can reach the desired functionality.